### DISCUSSION TITLE
Suggest a way of extracting critical JS that works with WDS

### DISCUSSION THREAD

**Discussion by Den-dp** [webpack/webpack#18216] [Q&A]
I have a pretty isolated JS file that is critical to run as early as possible (to reflect persisted theme preferences visually).

What is the best way to do this using webpack?

I've started with the simplest one:
```js
module.exports = {
  mode: "development",
  entry: {
    theme: "./src/theme.js",
    main: "./src/index.js",
  },
  plugins: [new HtmlWebpackPlugin()],
};
```
and it kind of works but breaks the WDS experience, inserting 2 WDS/HMR clients:

https://github.com/webpack/webpack/assets/1770529/7b044f0d-f227-4fba-826f-d9a0fbfc98c3

Here is the repo to play with https://github.com/Den-dp/webpack-critical-js-example

**Comment by alexander-akait** [comment]
You can just put `script` tag in your HTML file before our script tags

**Reply by Den-dp** [reply]
Right, but for that:
- I need to produce this file somehow (in reality `theme.js` is slightly more complex and requires `babel` and `typescript` loaders, so I want to keep using webpack for that)
- this js file must also not contain WDS Client (don't want to see 2 WDS error overlays)

Any ideas?

**Reply by alexander-akait** [reply]
How you run webpack dev server? You can disable all additinal entries, like hot and client, inject your entry with theme code firstly, and then inject hot and client again, there is not magic, just look at https://github.com/webpack/webpack-dev-server/blob/master/lib/Server.js#L510, maybe we can implement an option to make it easy

**Reply by Den-dp** [reply]
> How you run webpack dev server? 

I use [`webpack serve`](https://github.com/Den-dp/webpack-critical-js-example/blob/main/package.json#L7C15-L7C28) command and the following [`webpack.config.js`](https://github.com/Den-dp/webpack-critical-js-example/blob/main/webpack.config.js).

> You can disable all additinal entries...

Do you mean something like this https://github.com/Den-dp/webpack-critical-js-example/commit/f2882898e30bb088a84ca11465bff218ef475b43#diff-1fb26bc12ac780c7ad7325730ed09fc4c2c3d757c276c3dacc44bfe20faf166f ?

I wasn't happy with: 
* maintaining `"webpack-dev-server/client?hot=true&live-reload=true&overlay=true"` string by hand
* adding complexity to `webpack.config.js` via conditional plugins and entries (with and without WDS) depending on current target (build/serve).
* minor: it adds HMR-related code anyway (here is how [theme.js](https://github.com/Den-dp/webpack-critical-js-example/commit/f2882898e30bb088a84ca11465bff218ef475b43#diff-be0da604ab1105a43ece26c37d54752ea6078c3ffa186f7d4b54260472e254ba) looks in WDS mode)

https://github.com/webpack/webpack/assets/1770529/0b6afdf7-f7c2-44ec-9b39-5fe6b364ff4a

**Reply by alexander-akait** [reply]
You don't need https://github.com/Den-dp/webpack-critical-js-example/commit/f2882898e30bb088a84ca11465bff218ef475b43#diff-be0da604ab1105a43ece26c37d54752ea6078c3ffa186f7d4b54260472e254ba

Also please put your theme logic firstly here https://github.com/Den-dp/webpack-critical-js-example/commit/f2882898e30bb088a84ca11465bff218ef475b43#diff-1fb26bc12ac780c7ad7325730ed09fc4c2c3d757c276c3dacc44bfe20faf166fR9

So code will run before webpack-dev-server code

So you need something like:
```
main: [
  "./src/critical-code.js",
  "webpack-dev-server/client?hot=true&live-reload=true&overlay=true",
  "webpack/hot/dev-server",
  "./src/index.js",
],
```

**Reply by alexander-akait** [reply]
We need some code to run hot and reload logic, that is why you have delay, if you need real something fast you need to put it before any logic, and there are only two solution:
1. You can use multi compiler mode, i.e. **webpack.config.js** - `module.exports = [criticalCodeConfig, normalConfig];` and using https://webpack.js.org/configuration/other-options/#dependencies) run critical code compilation before normal then include this code in your HTML file before any other code
2. Modify dev server entries like above

The first solution is more flexible, because you can make more aggressive optimizations (look at terser options), avoid any extra async chunks (`ouptut.asyncChunks`) and more.
